africanus/calibration/utils.py

```diff
@@ -0,0 +1,332 @@
+                    t3*tmp[1, 1] +\
+                    t4*tmp[1, 1]
+    return njit(nogil=True)(subtract_model)
+
+
+@generated_jit(nopython=True, nogil=True, cache=True)
+def correct_vis(time_indices, antenna1, antenna2,
+                counts, jones, vis, flag):
+
+    mode = check_type(jones, vis)
+
+    jones_inverse_mul = jones_inverse_mul_factory(mode)
+
+    @wraps(correct_vis)
+    def _correct_vis_fn(time_indices, antenna1, antenna2,
+                        counts, jones, vis, flag):
+        jones_shape = np.shape(jones)
+        n_tim = jones_shape[0]
+        n_dir = jones_shape[3]
+        if n_dir > 1:
+            raise RuntimeError("Jones has n_dir > 1.\
+                                Cannot correct for direction dependent gains")
+        n_chan = jones_shape[2]
+        corrected_vis = np.zeros_like(vis, dtype=vis.dtype)
+        for t in range(n_tim):
+            for row in range(time_indices[t], time_indices[t] + counts[t]):
+                p = int(antenna1[row])
+                q = int(antenna2[row])
+                gp = jones[t, p]
+                gq = jones[t, q]
+                for nu in range(n_chan):
+                    if not np.any(flag[row, nu]):
+                        jones_inverse_mul(
+                            gp[nu, 0], vis[row, nu], gq[nu, 0],
+                            corrected_vis[row, nu])
+        return corrected_vis
+
+    return _correct_vis_fn
+
+
+@generated_jit(nopython=True, nogil=True, cache=True)
+def residual_vis(time_indices, antenna1, antenna2,
+                 counts, jones, vis, flag, model):
+
+    mode = check_type(jones, vis)
+    subtract_model = subtract_model_factory(mode)
+
+    @wraps(residual_vis)
+    def _residual_vis_fn(time_indices, antenna1, antenna2,
+                         counts, jones, vis, flag, model):
+        n_tim = np.shape(time_indices)[0]
+        vis_shape = np.shape(vis)
+        n_chan = vis_shape[1]
+        residual = np.zeros(vis_shape, dtype=vis.dtype)
+        for t in range(n_tim):
+            for row in range(time_indices[t], time_indices[t] + counts[t]):
+                p = int(antenna1[row])
+                q = int(antenna2[row])
+                gp = jones[t, p]
+                gq = jones[t, q]
+                for nu in range(n_chan):
+                    if not np.any(flag[row, nu]):
+                        subtract_model(
+                            gp[nu], vis[row, nu], gq[nu],
+                            model[row, nu], residual[row, nu])
+        return residual
+
+    return _residual_vis_fn
+
+
+RESIDUAL_VIS_DOCS = DocstringTemplate("""
+Computes residual visibilities in place
+given model visibilities and gains solutions.
+
+Parameters
+----------
+time_indices : $(array_type)
+    Time values of shape :code:`(row,)`
+antenna1 : $(array_type)
+    First antenna indices of shape :code:`(row,)`.
+antenna2 : $(array_type)
+    Second antenna indices of shape :code:`(row,)`
+counts : $(array_type)
```

Comment:
I think change the variable name to `time_counts` and move it just after `time_indices` so the time book-keeping variables are grouped together.